core/src/avm1/property.rs

```diff
@@ -17,115 +17,117 @@ bitflags! {
-        attributes: Attribute,
-    },
-    Stored {
-        value: Value<'gc>,
-        attributes: Attribute,
-    },
+pub struct Property<'gc> {
+    data: Option<Value<'gc>>,
+    getter: Option<Object<'gc>>,
+    setter: Option<Object<'gc>>,
+    attributes: Attribute,
 }
 
 impl<'gc> Property<'gc> {
-    /// Set a property slot.
+    pub fn new_stored(data: Value<'gc>, attributes: Attribute) -> Self {
+        Self {
+            data: Some(data),
+            getter: None,
+            setter: None,
+            attributes,
+        }
+    }
+
+    pub fn new_virtual(
+        getter: Object<'gc>,
+        setter: Option<Object<'gc>>,
+        attributes: Attribute,
+    ) -> Self {
+        Self {
+            data: None,
+            getter: Some(getter),
+            setter,
+            attributes,
+        }
+    }
+
+    pub fn data(&self) -> Option<Value<'gc>> {
+        self.data
+    }
+
+    pub fn getter(&self) -> Option<Object<'gc>> {
+        self.getter
+    }
+
+    pub fn setter(&self) -> Option<Object<'gc>> {
+        self.setter
+    }
+
+    /// Store data on this property, ignoring virtual setters.
     ///
-    /// This function may return an `Executable` of the property's virtual
-    /// function, if any happen to exist. It should be resolved, and its value
-    /// discarded.
-    pub fn set(&mut self, new_value: impl Into<Value<'gc>>) -> Option<Object<'gc>> {
-        match self {
-            Property::Virtual { set, .. } => {
-                if let Some(function) = set {
-                    Some(function.to_owned())
-                } else {
-                    None
-                }
-            }
-            Property::Stored {
-                value, attributes, ..
-            } => {
-                if !attributes.contains(Attribute::READ_ONLY) {
-                    *value = new_value.into();
-                }
-
-                None
+    /// Read-only properties are not affected.
+    pub fn set_data(&mut self, data: Value<'gc>) {
+        // Not using `is_overwritable` because virtual properties without a setter
+        // should be changed as well.
+        if !self.attributes.contains(Attribute::READ_ONLY) {
+            self.data = Some(data);
+        }
+    }
+
+    /// Make this property virtual by attaching a getter/setter to it.
+    pub fn set_virtual(&mut self, getter: Object<'gc>, setter: Option<Object<'gc>>) {
+        self.getter = Some(getter);
+        self.setter = setter;
+    }
+
+    /// Assign this property to a value.
+    ///
+    /// This function may return an `Object` of the property's virtual
+    /// setter, if it exists. It should be called by the caller.
+    pub fn set(&mut self, value: Value<'gc>) -> Option<Object<'gc>> {
+        if let Some(setter) = self.setter {
+            Some(setter)
+        } else {
+            if self.is_overwritable() {
+                self.data = Some(value);
             }
+            None
         }
     }
```

Comment:
I believe a set should always set the stored value, even if there is also a virtual property that hides it. E.g. even if we do:
```
var o = {};
o.addProperty("foo", get_foo, set_foo);
function get_foo() { trace("dummy"); return 0; }
function set_foo(n) { trace("dummy set"); }
o.foo = "bar";
```
the stored value of `foo` still gets set to `"bar"`. This can be observed in the Flash debugger (ctrl+shift+enter) where the value of the property will be displayed as "bar", or via other cases that return use `Object::get_data` to always return stored values. I think this happens in the case of watchers (cc #4827)

Reply:
Thanks, done. The data now always gets stored, but it's currently not visible to watchers (although it should be) because the `old_value` is currently fetched via a regular `TObject::get` that calls the getter, whereas it seems like it should be done by some sort of `ScriptObject::get_data` that looks up the prototype chain. But I think that should be fixed in a separate PR?